impl/Types.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+namespace BLA
```

Comment:
This file looks great. I agree it adds a bit of complexity to this library but it's really nice to be able to reason abut the DType of the Matrix when writing linear algebra functions. Great work!